kong/runloop/plugins_iterator.lua

```diff
@@ -123,86 +162,89 @@ local function get_next(self)
-        if plugin_configuration then
-          break
-        end
+    if service_id and consumer_id and combos[COMBO_SC] then
+      plugin_configuration = load_configuration(ctx, name, nil, service_id, consumer_id)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if route_id and service_id and combos[COMBO_RS] then
-        plugin_configuration = load_configuration(self, name, route_id, service_id, nil)
-        if plugin_configuration then
-          break
-        end
+    if route_id and service_id and combos[COMBO_RS] then
+      plugin_configuration = load_configuration(ctx, name, route_id, service_id, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if consumer_id and combos[COMBO_C] then
-        plugin_configuration = load_configuration(self, name, nil, nil, consumer_id)
-        if plugin_configuration then
-          break
-        end
+    if consumer_id and combos[COMBO_C] then
+      plugin_configuration = load_configuration(ctx, name, nil, nil, consumer_id)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if route_id and combos[COMBO_R] then
-        plugin_configuration = load_configuration(self, name, route_id, nil, nil)
-        if plugin_configuration then
-          break
-        end
+    if route_id and combos[COMBO_R] then
+      plugin_configuration = load_configuration(ctx, name, route_id, nil, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if service_id and combos[COMBO_S] then
-        plugin_configuration = load_configuration(self, name, nil, service_id, nil)
-        if plugin_configuration then
-          break
-        end
+    if service_id and combos[COMBO_S] then
+      plugin_configuration = load_configuration(ctx, name, nil, service_id, nil)
+      if plugin_configuration then
+        goto after_combos
       end
+    end
 
-      if combos[COMBO_GLOBAL] then
-        plugin_configuration = load_configuration(self, name, nil, nil, nil)
-      end
+    if combos[COMBO_GLOBAL] then
+      plugin_configuration = load_configuration(ctx, name, nil, nil, nil)
+    end
 
-    until true
+    ::after_combos::
 
     if plugin_configuration then
       ctx.plugins[name] = plugin_configuration
     end
   end
 
-  -- return the plugin configuration
-  if ctx.plugins[plugin.name] then
+  -- Return the plugin configuration if:
+  -- * The plugin must have a customized phase handler function (different from Base Plugin)
+  -- * The current phase is init_worker, (plugin name at least. config might be nil because no route_id/service_id etc in init_worker)
+  -- * A combo has been found and a conf has been loaded
+  local phase = self.iterator.phases[self.phase]
+  if phase and phase[plugin.name]
+  and (self.phase == "init_worker" or ctx.plugins[plugin.name]) then
```

Comment:
`init_worker` is such a special phase, only executed once on start on each worker, so I was thinking, could we remove that after the `init_worker` is done?

Reply:
You are right. I will switch the order in the statement.